Added switch, intentions section, proper rendering of credentials ++ (#47)
fredrikhgrelland/terraform-nomad-postgres: README.md

| use_host_volume | Use nomad host volume | bool | false | no |
| use_canary | Switch to use canary deployment for Postgres | bool | no |
| vault_secret.use_vault_provider | Set if want to access secrets from Vault | bool | true |
| vault_secret.vault_kv_path | Path to the secret key in Vault | string | "secret/postgres" |
| vault_secret.vault_kv_username_name | Secret key name in Vault kv path | string | "username" |
| vault_secret.vault_kv_password_name | Secret key name in Vault kv path | string | "password" |

## Outputs
| Name | Description | Type |
|------|-------------|------|
| service_name | Postgres service name | string |
| username | Postgres username | string |
| password | Postgres password | string |
| database_name | Postgres database name | string |
| port | Postgres port | number |

## Secrets & credentials
This module presents two ways of setting credentials (username and password). You can set them manually or upload secrets to Vault.

### Set credentials manually
Since this module uses Vault to store secrets by default, the `admin_user` and `admin_password` is not used. 

> However, if you don't want to use Vault in your application or just want to change them, they are easy to configure and costumize.

Reviewer: ```suggestion
However, if you don't want to use Vault in your application or just want to change them, they are easy to configure and customize.
```

> All you need to do is set the `vault_secret.use_vault_provider` to `false` and leave the rest of the values as blanks. Finally, you can choose to change the default username and password. E.g:

Reviewer: IMHO,
`default` should be false. 
User-provided credentials are the default settings for the module.

Author: I can agree to that - will fix!

                  }
  admin_user     = "myadminuser"     # default 'postgres'
  admin_password = "myadminpassword" # default 'postgres'
}
```

### Set credentials using Vault secrets
By default the postgres username and password is generated and put in `secret/postgres` inside Vault, using the [01_generate_secrets_vault.yml](dev/ansible/01_generate_secrets_vault.yml) playbook.

If you where to change the secrets path and keys/values in Vault with your own configuration you would need to change the variables in the `vault_secret`-object. 
Say that you have put your secrets in `secret/services/postgres/users` and change the keys to `guestuser` and `guestpassword`. Then you need to do the following configuration:
```hcl-terraform
module "postgres" {
...

  vault_secret  = {
                    use_vault_provider     = true,
                    vault_kv_path          = "secret/services/postgres/users",
                    vault_kv_username_name = "guestuser",
                    vault_kv_password_name = "guestpassword"
                  }
}
```

## Volumes
We are using [host volume](https://www.nomadproject.io/docs/job-specification/volume) to store postgres data.

> If the `use_host_volume` is set to `true`, Postgres data will be available in root `/persistence/postgres` folder inside the Vagrant box.

Reviewer: ```suggestion
If the `use_host_volume` is set to `true` (default: false), Postgres data will be available in root `/persistence/postgres` folder inside the Vagrant box.
```

## Content
1. [Requirements](#requirements)
    1. [Required software](#required-software)
2. [Usage](#usage)
    1. [Providers](#providers)
    2. [Intentions](#intentions)
3. [Inputs](#inputs)
4. [Outputs](#outputs)
5. [Secrets & credentials](#secrets--credentials)
6. [Volumes](#volumes)
7. [Example usage](#example-usage)
    1. [Verifying setup](#verifying-setup)
8. [License](#license)

## Requirements
### Required software
- [GNU make](https://man7.org/linux/man-pages/man1/make.1.html)
- [Docker](https://www.docker.com/)

## Usage
```text
make test
```
The command will run a standalone instance of postgres found in the [example](/example) folder.

### Providers
- [Nomad](https://registry.terraform.io/providers/hashicorp/nomad/latest/docs)

### Intentions
The following intentions is created in the Ansible playboook [00_create_intetion.yml](dev/ansible/00_create_intention.yml):

| Intention between | type |
| :---------------- | :--- |
| postgres-local => postgres  | allow |

> :warning: Note that these intentions needs to be created if you are using the module in another module.

> 

Reviewer 2: I don't think we should include information like the intentions created by ansible-playbooks in this README. I believe this README should solely be for the postgres module, and the ansible-playbooks for example are actually a part of the testing inside a vagrantbox. I think this should be rephrased as a "these intentions will need to be put in place if you are going to use this module in a hashistack ecosystem, we have done so in our vagrantbox example".
Any thoughts?

Author: That's a good suggestion. I'll do that!

| use_host_volume | Use nomad host volume | bool | false | no |
| use_canary | Switch to use canary deployment for Postgres | bool | no |
| vault_secret.use_vault_kv | Set if want to access secrets from Vault | bool | false |
| vault_secret.vault_kv_path | Path to the secret key in Vault | string | "secret/postgres" |
| vault_secret.vault_kv_username_name | Secret key name in Vault kv path | string | "username" |
| vault_secret.vault_kv_password_name | Secret key name in Vault kv path | string | "password" |

## Outputs
| Name | Description | Type |
|------|-------------|------|
| service_name | Postgres service name | string |
| username | Postgres username | string |
| password | Postgres password | string |
| database_name | Postgres database name | string |
| port | Postgres port | number |

## Secrets & credentials
This module presents two ways of setting credentials (username and password). You can set them manually or upload secrets to Vault.

### Set credentials manually
This module uses the `admin_user` and `admin_password` credentials as default, and contain the value `postgres`.  

> If you want to overwrite the default values, you need to is set the `vault_secret.use_vault_kv` to `false` and leave the rest of the values as blanks. 

Reviewer: I am a bit lost in this sentence. Do you mean override default values of `admin_user` and `admin_password`?
```suggestion
If you want to overwrite the default values, you need to set the `vault_secret.use_vault_kv` to `false` and leave the rest of the values as blanks. 
```

Author: Yes, that's right. It's only ment to indicate that you can overwrite the default values that is set to `postgres`. Do you want me to rephrase the sentence?

Reviewer: Yes

> ### Set credentials manually
> This module uses the `admin_user` and `admin_password` credentials as default, and contain the value `postgres`.  
> If you want to overwrite the default values, you need to is set the `vault_secret.use_vault_kv` to `false` and leave the rest of the values as blanks. 
> Finally, you can change the default username and password. 

Reviewer 2: ```suggestion
### Set credentials manually
To set the credentials manually you first need to tell the module to not fetch credentials from vault. To do that set `vault_secret.use_vault_kv` to `false` (see below for example). If this is done the module will use the variables `admin_user` and `admin_password` to set the postgres credentials. These will default to `postgres` if not set by the user.  
Below is an example on how to disable the use of vault credentials, and setting your own credentials.
```

Author: Done!

E.g:
```hcl-terraform

> module "postgres" {

Reviewer 2: Add a small hint to what file this is in
```suggestion
`main.tf`
```hcl-terraform
module "postgres" {
```

Author: I'll rephrase to something similar 😄

...

  vault_secret  = {
                    use_vault_kv           = false,
                    vault_kv_path          = "",
                    vault_kv_username_name = "",
                    vault_kv_password_name = ""
                  }
  admin_user     = "myadminuser"     # default 'postgres'
  admin_password = "myadminpassword" # default 'postgres'
}
```

### Set credentials using Vault secrets

> By default the Vault secrets are not activated. In other words the `vault_secret.use_vault_kv` is set to `false`. 

Reviewer 2: ```suggestion
By default `use_vault_kv` is set to `false`. 
```

Author: 😅

However, when testing using the box (e.g. `make dev`) the postgres username and password is randomly generated and put in `secret/postgres` inside Vault, from the [01_generate_secrets_vault.yml](dev/ansible/01_generate_secrets_vault.yml) playbook. 

> This is a independet process and will run regardless of the `vault_secret.use_vault_kv` is `false/true`. 

Reviewer 2: ```suggestion
This is an independet process and will run regardless of the `vault_secret.use_vault_kv` is `false/true`. 
```

> If you want to use the automatically generated credentials in the box, you can do so by changing the `vault_secret` objcet as seen bellow:

Reviewer: ```suggestion
If you want to use the automatically generated credentials in the box, you can do so by changing the `vault_secret` object as seen below:
```

```hcl-terraform
module "postgres" {
...

  vault_secret  = {
                    use_vault_kv           = true,
                    vault_kv_path          = "secret/postgres",
                    vault_kv_username_name = "username",
                    vault_kv_password_name = "password"
                  }
}
```

> If you where to change the secrets path and keys/values in Vault with your own configuration you would need to change the variables in the `vault_secret`-object. 

Reviewer: ```suggestion
If you want to change the secrets path and keys/values in Vault with your own configuration you would need to change the variables in the `vault_secret`-object. 
```

Say that you have put your secrets in `secret/services/postgres/users` and change the keys to `guestuser` and `guestpassword`. Then you need to do the following configuration:
```hcl-terraform
module "postgres" {
...

  vault_secret  = {
                    use_vault_kv           = true,
                    vault_kv_path          = "secret/services/postgres/users",
                    vault_kv_username_name = "guestuser",
                    vault_kv_password_name = "guestpassword"
                  }
}
```

## Volumes

> We are using [host volume](https://www.nomadproject.io/docs/job-specification/volume) to store postgres data.

Reviewer: ```suggestion
Module (optionally) supports [host volume](https://www.nomadproject.io/docs/job-specification/volume) to store postgres data.
```

If the `use_host_volume` is set to `true` (default: false), Postgres data will be available in root `/persistence/postgres` folder inside the Vagrant box.

## Example usage

> The example-code shows the minimum of what you need to use this module.

Reviewer 2: Technically none of the variables are needed as a minimum, because they have defaults. But that might not be as instructive in learning how to use the module maybe?

What about removing all variables from the minimum example, and provide a link to where you can see the defaults + an example on how to override them

Author: I'll rephrase it!:+1:
